Interp points on element

markscheel opened the pull request:
## Proposed changes

This PR does a few things:

 - Refactors more of the interpolation code
 - Adds new functions for those InterpolationTargets that have time-independent points.
 - Adds 3 Actions associated with putting interpolation points into DgElementArray
 - Adds a test for the 3 new actions

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Add time-indep points() functions to InterpolationTargets.

Also, don't pass ConstGlobalCache anymore since it is no
longer needed. Instead, pass tmpl::type_<Metavariables> so
that Metavariables can be deduced.

markscheel committed:
Refactor AddTemporalIdsToInterpolationTarget.

This just moves some code from an Action into
a function that will eventually be called from more
than one Action.

markscheel committed:
Add new Tag holding interpolation points on Element.

markscheel committed:
Add Test_ElementReceiveInterpPoints.cpp

markscheel committed:
Add ElementInitInterpPoints.

markscheel committed:
Add InterpolationTargetSendPoints.

markscheel committed:
Add ElementReceiveInterpPoints

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -268,6 +268,57 @@ bool have_data_at_all_points(const db::DataBox<DbTags>& box,
+/// of InterpolationTarget
+///
+/// Currently two Actions call flag_temporal_ids_for_interpolation:
+///   AddTemporalIdsToInterpolationTarget (called by Events::Interpolate)
```
add `-` to render the two items in a list

kidder replied:
it looks like you fixed this in another commit?

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp:
```diff
@@ -171,6 +171,12 @@ struct KerrHorizon {
     return db::get<StrahlkorperTags::CartesianCoords<::Frame::Inertial>>(box);
   }
+  template <typename Metavariables, typename DbTags>
+  static auto points(
```
auto -> known type

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetLineSegment.hpp:
```diff
@@ -148,6 +147,14 @@ struct LineSegment {
+      const db::DataBox<DbTags>& box,
+      Parallel::ConstGlobalCache<Metavariables>& cache,
+      const TemporalId& /*temporal_id*/) noexcept {
+    return points(box, cache);
```
is the copy intentional?

markscheel replied:
Yes, it returns an object that was just computed, not a const reference.

nilsdeppe replied:
Sounds good!

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetLineSegment.hpp:
```diff
@@ -126,11 +126,10 @@ struct LineSegment {
       const db::DataBox<DbTags>& box,
-      Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
-      const TemporalId& /*temporal_id*/) noexcept {
+      Parallel::ConstGlobalCache<Metavariables>& /*cache*/) noexcept {
```
const

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetKerrHorizon.hpp:
```diff
@@ -171,6 +171,12 @@ struct KerrHorizon {
+  template <typename Metavariables, typename DbTags>
+  static auto points(
+      const db::DataBox<DbTags>& box,
+      Parallel::ConstGlobalCache<Metavariables>& /*cache*/) noexcept {
```
const

kidder replied:
is the cache used for anything?  if not, don't pass it in...

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetLineSegment.hpp:
```diff
@@ -148,6 +147,14 @@ struct LineSegment {
+  template <typename Metavariables, typename DbTags, typename TemporalId>
+  static tnsr::I<DataVector, VolumeDim, Frame::Inertial> points(
+      const db::DataBox<DbTags>& box,
+      Parallel::ConstGlobalCache<Metavariables>& cache,
```
const

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -194,11 +194,10 @@ struct WedgeSectionTorus {
       const db::DataBox<DbTags>& box,
-      Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
-      const TemporalId& /*temporal_id*/) noexcept {
+      Parallel::ConstGlobalCache<Metavariables>& /*cache*/) noexcept {
```
const

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetWedgeSectionTorus.hpp:
```diff
@@ -284,6 +283,14 @@ struct WedgeSectionTorus {
+  template <typename Metavariables, typename DbTags, typename TemporalId>
+  static tnsr::I<DataVector, 3, Frame::Inertial> points(
+      const db::DataBox<DbTags>& box,
+      Parallel::ConstGlobalCache<Metavariables>& cache,
```
const

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -394,6 +394,8 @@ void add_received_variables(
 ///                             and by FindApparentHorizon)
 /// - InterpolationTargetVarsFromElement (called by DgElementArray)
+/// - InterpolationTargetSendTimeIndepPointsToElements
+///  (in InterpolationTarget ActionList)
```
indent one more char

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -406,6 +408,20 @@ auto get_block_logical_coords(const db::DataBox<DbTags>& box,
+          typename Metavariables>
+auto get_block_logical_coords(
+    const db::DataBox<DbTags>& box,
+    Parallel::ConstGlobalCache<Metavariables>& cache) noexcept {
```
const and is it used?

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp:
```diff
@@ -406,6 +408,20 @@ auto get_block_logical_coords(const db::DataBox<DbTags>& box,
+/// are time-independent.
+template <typename InterpolationTargetTag, typename DbTags,
+          typename Metavariables>
+auto get_block_logical_coords(
```
why not just name it `block_logical_coords`?

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolatedVars.hpp:
```diff
@@ -115,5 +115,23 @@ struct HolderTag {
+/// is accessed from the `Element`'s `DataBox` with tag
+/// `Tags::InterpPointInfo`.
+///
+/// This is for the case in which interpolation bypasses the
```
This tag is

kidder replied:
can this be put in a separate header to reduce dependencies...

markscheel replied:
I'm not sure how you want to reduce dependencies.  Do you want `Tags::InterpPointInfo` and `Vars::PointInfoTag` to be both put into a new header, e.g. PointInfoTag.hpp ?

nilsdeppe replied:
Well it looks like `PointInfoTag` doesn't depend on anything else in this file, and is what is used in the element, so just moving that to a separate file would reduce the amount of unnecessary dependencies, right?

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolationTargetSendPoints.hpp:
```diff
@@ -0,0 +1,55 @@
+/// \ingroup ActionsGroup
+/// \brief Sends interpolation points to all the Elements.
+///
+/// This is for the case in which the points are time-independent.
```
This action is

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/InterpolationTargetSendPoints.hpp:
```diff
@@ -0,0 +1,55 @@
+///
+/// This is for the case in which the points are time-independent.
+///
+/// This should be placed in the Registration PDAL for InterpolationTarget.
```
This action should

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -0,0 +1,75 @@
+/// \ingroup ActionsGroup
+/// \brief Adds interpolation point holders to the Element's DataBox.
+///
+/// This is for the case in which the points are time-independent.
```
This action

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -0,0 +1,75 @@
+///
+/// This is for the case in which the points are time-independent.
+///
+/// This should be placed in the Initialization PDAL for DgElementArray.
```
This action

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -0,0 +1,75 @@
+            db::RemoveTags<>,
+            db::AddSimpleTags<intrp::Tags::InterpPointInfo<Metavariables>>>(
+            std::move(box),
+            db::item_type<intrp::Tags::InterpPointInfo<Metavariables>>{}));
```
use the known type

markscheel replied:
In this case the type is `tuples::tagged_tuple_from_typelist<
      db::wrap_tags_in<point_info_detail::WrappedPointInfoTag,
                       typename Metavariables::interpolation_target_tags,
                       Metavariables>>` which has a lot of stuff going on in it.  Should I still use that explicit type here?  (I guess I'm not sure about under what circumstances it is ok to use db::item_type).

Or I can call it `Tags::InterpPointInfo<Metavariables>::type` ...

nilsdeppe replied:
My vote would be for a type alias to the long thing, and then using the type alias inside. We're trying to avoid `item_type`-style code because it makes the code as difficult to read as using `auto` everywhere (I find, at least).

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -0,0 +1,75 @@
+            db::item_type<intrp::Tags::InterpPointInfo<Metavariables>>{}));
+  }
+
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
```
please combine these with if constexpr. (see #2200) and in the second case `ERROR`

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_ElementReceiveInterpPoints.cpp:
```diff
@@ -0,0 +1,228 @@
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <boost/optional/optional_io.hpp>
```
alphabetical

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_ElementReceiveInterpPoints.cpp:
```diff
@@ -0,0 +1,228 @@
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Initialization,
+          tmpl::list<intrp::Actions::ElementInitInterpPoints>>,
+      Parallel::PhaseActions<typename Metavariables::Phase,
```
don't need to add empty list to PDAL as long as one phase is listed

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_ElementReceiveInterpPoints.cpp:
```diff
@@ -0,0 +1,228 @@
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting::MockArrayChare;
+  using array_index = size_t;
+  using const_global_cache_tags = tmpl::flatten<tmpl::append<
```
can't you just use tmpl::push_back?

kidder reviewed (changes requested):
reviewed with @nilsdeppe , you can squash the fixups

markscheel commented:
Made all changes, squashed, rebased onto develop (to get the cpp17 stuff).

kidder commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_ElementReceiveInterpPoints.cpp:
```diff
@@ -0,0 +1,224 @@
+  using const_global_cache_tags = tmpl::push_back<
+      Parallel::get_const_global_cache_tags_from_actions<
+          tmpl::list<typename InterpolationTargetTag::compute_target_points>>,
+      tmpl::list<domain::Tags::Domain<Metavariables::volume_dim>>>;
```
just push the tag not a one-element list

kidder reviewed (changes requested):
you can squash the two minor comments

markscheel committed:
InterpolationTargetDetail: Add time-indep block_logical_coords.

Also, rename function to block_logical_coordinates
instead of get_block_logical_coordinates and don't pass
ConstGlobalCache as an argument because it is no longer needed.
(these last two changes required changes to SendPointsToInterpolator).

markscheel commented:
made changes and squashed.

nilsdeppe commented:
Restarted CI since it looks like GitHub has fixed the reduced space issue that caused CI to die.
